satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,467 @@
+		err = stream.Send(completed)
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitInitiatedAt == nil {
+		request := &overlay.ExitStatusRequest{NodeID: nodeID, ExitInitiatedAt: time.Now().UTC()}
+		_, err = endpoint.overlaydb.UpdateExitStatus(ctx, request)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = stream.Send(&pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}})
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitLoopCompletedAt == nil {
+		err = stream.Send(&pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}})
+		return Error.Wrap(err)
+	}
+
+	pending := newPendingMap()
+
+	var morePiecesFlag int32 = 1
+
+	var group errgroup.Group
+	group.Go(func() error {
+		ticker := time.NewTicker(endpoint.interval)
+		defer ticker.Stop()
+
+		for range ticker.C {
+			if pending.length() == 0 {
+				incomplete, err := endpoint.db.GetIncompleteNotFailed(ctx, nodeID, endpoint.config.EndpointBatchSize, 0)
+				if err != nil {
+					return Error.Wrap(err)
+				}
+
+				if len(incomplete) == 0 {
+					incomplete, err = endpoint.db.GetIncompleteFailed(ctx, nodeID, endpoint.config.EndpointMaxFailures, endpoint.config.EndpointBatchSize, 0)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+				}
+
+				if len(incomplete) == 0 {
+					endpoint.log.Debug("no more pieces to transfer for node.", zap.String("node ID", nodeID.String()))
+					atomic.StoreInt32(&morePiecesFlag, 0)
+					break
+				}
+
+				for _, inc := range incomplete {
+					err = endpoint.processIncomplete(ctx, stream, pending, nodeID, inc)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+				}
+			}
+		}
+		return nil
+	})
+
+	for {
+		pendingCount := pending.length()
+		// if there are no more transfers and the pending queue is empty, send complete
+		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingCount == 0 {
+			// TODO check whether failure threshold is met before sending completed
+			// TODO needs exit signature
+			transferMsg := &pb.SatelliteMessage{
+				Message: &pb.SatelliteMessage_ExitCompleted{
+					ExitCompleted: &pb.ExitCompleted{},
+				},
+			}
+			err = stream.Send(transferMsg)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			break
+		}
+		// skip if there are none pending
+		if pendingCount == 0 {
+			continue
+		}
+
+		request, err := stream.Recv()
+		if err != nil {
+			return eofHandler(err)
+		}
+
+		switch m := request.GetMessage().(type) {
+		case *pb.StorageNodeMessage_Succeeded:
+			err = endpoint.handleSucceeded(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
+		case *pb.StorageNodeMessage_Failed:
+			err = endpoint.handleFailed(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+		default:
+			return Error.New("unknown storage node message: %v", m)
+		}
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Would it be enough just do `return group.Wait()` here?

Reply:
I think this would be fine.